plugins/renepay/pay.c

```diff
@@ -0,0 +1,1798 @@
+		json_add_string(ret, "status", "complete");
+		json_add_amount_msat(ret, "amount_msat", complete_msat);
+		json_add_amount_msat(ret, "amount_sent_msat",complete_sent);
+		json_add_node_id(ret, "destination", &p->destination);
+		json_add_sha256(ret, "payment_hash", &p->payment_hash);
+		json_add_u32(ret, "created_at", complete_created_at);
+		json_add_num(ret, "parts", complete_parts);
+
+		/* This payment was already completed, we don't keep record of
+		 * it twice. */
+		renepay->payment = tal_free(renepay->payment);
+
+		return command_finished(cmd, ret);
+	} else if (pending) {
+		p->groupid = last_pending_group_id;
+		renepay->next_partid = last_pending_partid+1;
+
+		p->total_sent = pending_sent;
+		p->total_delivering = pending_msat;
+
+		plugin_log(pay_plugin->plugin,LOG_DBG,
+			   "There are pending sendpays to this invoice. "
+			   "groupid = %ld, "
+			   "delivering = %s, "
+			   "last_partid = %ld",
+			   last_pending_group_id,
+			   type_to_string(tmpctx,struct amount_msat,&p->total_delivering),
+			   last_pending_partid);
+
+		if( first_pending_group_id != last_pending_group_id)
+		{
+			/* At least two pending groups for the same invoice,
+			 * this is weird, we better stop. */
+			renepay->payment = tal_free(renepay->payment);
+			return renepay_fail(renepay, PAY_IN_PROGRESS,
+					    "Payment is pending by some other request.");
+		}
+		if(amount_msat_greater_eq(p->total_delivering,p->amount))
+		{
+			/* Pending payment already pays the full amount, we
+			 * better stop. */
+			renepay->payment = tal_free(renepay->payment);
+			return renepay_fail(renepay, PAY_IN_PROGRESS,
+					    "Payment is pending with full amount already commited");
+		}
+	}else
+	{
+		p->groupid = (last_group==INVALID_ID  ? 1 : (last_group+1)) ;
+		renepay->next_partid=1;
+	}
+
+
+	struct out_req *req;
+	/* Get local capacities... */
+	req = jsonrpc_request_start(cmd->plugin, cmd, "listpeerchannels",
+				    listpeerchannels_done,
+				    listpeerchannels_done, renepay);
+	return send_outreq(cmd->plugin, req);
+}
+
+static struct command_result *json_pay(struct command *cmd,
+				       const char *buf,
+				       const jsmntok_t *params)
+{
+	const char *invstr;
+	const char *label;
+	const char *description;
+	struct sha256 * local_offer_id;
+ 	u64 invexpiry;
+ 	struct amount_msat *msat, *invmsat;
+	struct amount_msat *maxfee;
+	u64 *riskfactor_millionths;
+	u32 *maxdelay;
+	u64 *base_fee_penalty;
+	u64 *prob_cost_factor;
+	u64 *min_prob_success_millionths;
+	u32 *retryfor;
+
+#if DEVELOPER
+	bool *use_shadow;
+#endif
+
+	if (!param(cmd, buf, params,
+		   p_req("invstring", param_string, &invstr),
+ 		   p_opt("amount_msat", param_msat, &msat),
+ 		   p_opt("maxfee", param_msat, &maxfee),
+
+		   // MCF parameters
+		   // TODO(eduardo): are these parameters read correctly?
+		   p_opt_def("base_fee_penalty", param_millionths, &base_fee_penalty,10),
+ 		   p_opt_def("prob_cost_factor", param_millionths, &prob_cost_factor,10),
```

Comment:
I always prefer fewer parameters, and our users do too!  We can always *add* more later, anyway.

What has turned out to make sense is "here's the maximum fee to pay".  Then you do the Right Thing:
1. If you can't find any route with that low a fee, fail and tell them the fee they'd need.
2. If you can find a (vastly) unlikely route with that fee, try it (once) before giving up, *even* if you normally wouldn't bother.  Also, tell them what a reasonable fee looks like if it does fail.  
3. Otherwise, try until you time out.

But whenever you fail, try to indicate a meta-level cause for failure: 
- Was there a higher fee which would help? 
- Did it take so long for a single payment to time out that you only got one attempt?  
- Was there simply no likely route at any price?
- Was there some critical channel which was unavailable?  e.g.  Alice paying Bob, but the direct channel is down, or was there an obvious "best" incoming to Bob which proved not to work?

*actionable* feedback wins!  In this case:
1. increase maxfee
2. increase timeout
3. open a new channel
4. complain on nostr about Bob

Reply:
This parameter `prob_cost_factor` is allowing us to linearly combine probability cost and fees, which are two completely different things. It's like a units conversion. However, it is not a fixing how much are interested in prob. with respect to fees, because the cost function contains another scalar `mu` :
```
cost = prob_cost*prob_cost_factor*mu + fees*(MAX_MU-mu)
```
the smaller the `mu` then fees are more important, while the bigger the value of `mu` then we care more about probabilities. Then `mu` is selected by the algorithm that tries to find a low cost flow that satisfies the user constraints on reliability and fee budget. 
So `prob_cost_factor` does not determine the priority, but it gives a prior advantage to the probability cost.

If you don't care about reliability you can set it to zero, but if you care more about probabilities than fees then you choose it as a very large number.

I agree with Rusty saying fewer parameters is better, but we haven't explored what is the best value for this `prob_cost_factor`, so maybe 10 is a terrible choice to be default.

Reply:
But it's a derived parameter: user cares about fees and time.

Sure, the derivation is complex :) But that's also a very good argument against asking the user to set it!

Reply:
Should I also remove base fee penalty as an option? That one is also not very user-friendly.

Reply:
I removed `prob_cost_factor` as an option, now it only has a default value, but I can still be changed in developer mode. https://github.com/ElementsProject/lightning/pull/6376/commits/dc8071cf32be6e8d9bb99e15937e2755f256efd4